File: leetcode/ds/3043. Find the Length of the Longest Common Prefix/3043. Find the Length of the Longest Common Prefix.cpp

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEndOfWord;
    
    TrieNode() {
        isEndOfWord = false;
    }
};

class Trie {
private:
    TrieNode* root;

public:
    Trie() {
        root = new TrieNode();
    }

    void insert(const string& word) {
        TrieNode* node = root;
        for (char ch : word) {
            if (node->children.find(ch) == node->children.end()) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
        }
        node->isEndOfWord = true;
    }

    int findCommonPrefixLength(const string& word) {
        TrieNode* node = root;
        int prefixLen = 0;
        for (char ch : word) {
            if (node->children.find(ch) == node->children.end()) {
                break;
            }
            node = node->children[ch];
            prefixLen++;
        }
        return prefixLen;
    }
};

class Solution {
public:
    int longestCommonPrefix(const vector<int>& arr1, const vector<int>& arr2) {
        Trie trie;
        for (int num : arr2) {
            trie.insert(to_string(num));
        }

        int longestCommonLength = 0;

        for (int num : arr1) {
            string numStr = to_string(num);
            int commonPrefixLength = trie.findCommonPrefixLength(numStr);
            longestCommonLength = max(longestCommonLength, commonPrefixLength);
        }

        return longestCommonLength;
    }
};

File: leetcode/ds/3043. Find the Length of the Longest Common Prefix.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEndOfWord = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.isEndOfWord = True

    def findCommonPrefixLength(self, word):
        node = self.root
        prefixLen = 0
        for char in word:
            if char not in node.children:
                break
            node = node.children[char]
            prefixLen += 1
        return prefixLen
    
class Solution(object):
    def longestCommonPrefix(self, arr1, arr2):
        trie = Trie()
        for num in arr2:
            trie.insert(str(num))

        longestCommonLength = 0
        
        for num in arr1:
            numStr = str(num)
            commonPrefixLength = trie.findCommonPrefixLength(numStr)
            longestCommonLength = max(longestCommonLength, commonPrefixLength)

        return longestCommonLength
